crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,359 @@
+use std::{
+    env,
+    ffi::OsStr,
+    fs,
+    io::{BufReader, Cursor, ErrorKind, Read},
+    path::Path,
+};
+
+use semver::Version;
+use serde::Deserialize;
+use typst::{diag::bail, diag::StrResult, eval::eco_format};
+
+use crate::args::UpdateCommand;
+
+// these might not be very usefull but does make it easier to maintain
+// if the organization/repo moves or changes (only used in release fetching)
+const TYPST_GITHUB_ORG: &str = "typst";
+const TYPST_REPO: &str = "typst";
+
+#[cfg(feature = "no-self-update")]
+pub(crate) const NEVER_SELF_UPDATE: bool = true;
+#[cfg(not(feature = "no-self-update"))]
+pub(crate) const NEVER_SELF_UPDATE: bool = false;
+
+/// Figure out if there are sufficient permissions to carry out an update --
+/// hard failing if the cli is installed through a package manager.
+#[derive(Clone, Copy, Debug)]
+enum SelfUpdatePermission {
+    HardFail,
+    Permit,
+}
+
+/// A GitHub release.
+#[derive(Debug, Deserialize)]
+struct Release {
+    name: String,
+    tag_name: String,
+    assets: Vec<Asset>,
+}
+
+/// Assets that were uploaded to a GitHub release.
+///
+/// Primarly used to download pre-compiled typst CLI binaries.
+#[derive(Debug, Deserialize)]
+struct Asset {
+    pub name: String,
+    pub browser_download_url: String,
+}
+
+/// GitHub asset archive with the typst CLI executable as unpacked data.
+#[derive(Debug)]
+struct Archive {
+    pub extension: Extension,
+    pub buffer: BufReader<Cursor<Vec<u8>>>,
+}
+
+/// The extension for the downloaded archive.
+///
+/// Possible variations are `zip` and `xz`, other variations are mapped to
+/// `Unsupported` and will throw an error.
+#[derive(Debug)]
+enum Extension {
+    Zip,
+    Xz,
+    Unsupported,
+}
+
+impl From<&str> for Extension {
+    fn from(value: &str) -> Self {
+        match value {
+            "zip" => Extension::Zip,
+            "xz" => Extension::Xz,
+            _ => Extension::Unsupported,
+        }
+    }
+}
+
+/// Self update the typst CLI binary.
+///
+/// Fetches a target release or the latest release (if no version was specified)
+/// from GitHub, unpacks it and self replaces the current binary with the
+/// pre-compiled asset from the downloaded release.
+pub fn update(command: UpdateCommand) -> StrResult<()> {
+    let update_permitted = if NEVER_SELF_UPDATE {
+        SelfUpdatePermission::HardFail
+    } else {
+        self_update_permitted()?
+    };
+
+    match update_permitted {
+        SelfUpdatePermission::HardFail => {
+            println!("self-update is disabled for this build of the typst cli");
+            println!(
+                "you should probably use your system package manager to update typst"
+            );
+            // not really an ok scenario but not really an error either?
+            return Ok(());
```

Comment:
Does this cause the command to return a non-zero exit code? (This is important to indicate the update failed.)
In general, if the user tried to update but it did not occur, the command should return a non-zero exit code.

Could also be more appropriate to use `eprintln!` here, but I'm not particularly sure about that in specific.